### ISSUE TITLE
failed to compile perl5  when path includes the string "blib"

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14801]
Migrated from [rt.perl.org#125603](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=125603) (status was 'resolved')

Searchable as RT125603$

**Comment by p5pRT** [comment]
### From rongqing.li@windriver.com

1\. failed to compile perl5  when path includes the string "blib"
Reproduce as below&#8203;:

\#mkdir blib
\#cd blib
\#unxz perl\-5\.22\.0\.tar\.xz
\#tar \-xvf perl\-5\.22\.0\.tar
\#cd perl\-5\.22\.0
\#\./Configure \-des \-Dprefix=\./
\#make


2\. the possible patch is below&#8203;:
\+\-\-\-
\+ dist/PathTools/Makefile\.PL | 3 \-\-\-
\+ 1 file changed\, 3 deletions\(\-\)
\+
\+diff \-\-git a/dist/PathTools/Makefile\.PL b/dist/PathTools/Makefile\.PL
\+index 1b21de4\.\.f562cb2 100644
\+\-\-\- a/dist/PathTools/Makefile\.PL
\+\+\+\+ b/dist/PathTools/Makefile\.PL
\+@&#8203;@&#8203; \-1\,6 \+1\,3 @&#8203;@&#8203;
\+\-
\+\-BEGIN \{ @&#8203;INC = grep \{\!/blib/\} @&#8203;INC \}
\+\-
\+ require 5\.005;
\+ use ExtUtils&#8203;::MakeMaker;
\+ WriteMakefile
\+\-\- 
\+1\.9\.1

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Jul 12 19&#8203;:08&#8203;:58 2015\, rongqing\.li@&#8203;windriver\.com wrote&#8203;:
> 1\. failed to compile perl5  when path includes the string "blib"
> Reproduce as below&#8203;:
> 
> \#mkdir blib
> \#cd blib
> \#unxz perl\-5\.22\.0\.tar\.xz
> \#tar \-xvf perl\-5\.22\.0\.tar
> \#cd perl\-5\.22\.0
> \#\./Configure \-des \-Dprefix=\./
> \#make
> 
> 
> 2\. the possible patch is below&#8203;:
> \+\-\-\-
> \+ dist/PathTools/Makefile\.PL | 3 \-\-\-
> \+ 1 file changed\, 3 deletions\(\-\)
> \+
> \+diff \-\-git a/dist/PathTools/Makefile\.PL b/dist/PathTools/Makefile\.PL
> \+index 1b21de4\.\.f562cb2 100644
> \+\-\-\- a/dist/PathTools/Makefile\.PL
> \+\+\+\+ b/dist/PathTools/Makefile\.PL
> \+@&#8203;@&#8203; \-1\,6 \+1\,3 @&#8203;@&#8203;
> \+\-
> \+\-BEGIN \{ @&#8203;INC = grep \{\!/blib/\} @&#8203;INC \}
> \+\-
> \+ require 5\.005;
> \+ use ExtUtils&#8203;::MakeMaker;
> \+ WriteMakefile
> \+\-\- 
> \+1\.9\.1

PathTools is a library which is distributed with the Perl 5 core\, but also available on CPAN \(http&#8203;://search\.cpan\.org/dist/PathTools/\)\.  The fact that it is in the 'dist/' means that it is primarily maintained by the Perl 5 Porters\.

However\, any deficiency in dist/PathTools/Makefile\.PL would \*at most\* affect the compilation of the PathTools library\.  It would \*not\* affect the compilation of perl 5 as a whole\.  So I doubt that a change to that Makefile\.PL would be the remedy to your problem \-\- if indeed your problem is a failure of perl5 as a whole?

Could you please post\, as an attachment to this bug report \(or email attachment\)\, the complete arguments you used to configure Perl and all the output of 'sh \./Configure' and 'make' up to the point where you experienced a failure?

Also\, please let us know your OS; how you are attempting to build Perl 5; etc\.

Thank you very much\.
\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From rongqing.li@windriver.com

On Mon Jul 13 15&#8203;:37&#8203;:54 2015\, jkeenan wrote&#8203;:
> On Sun Jul 12 19&#8203;:08&#8203;:58 2015\, rongqing\.li@&#8203;windriver\.com wrote&#8203;:
> > 1\. failed to compile perl5  when path includes the string "blib"
> > Reproduce as below&#8203;:
> >
> > \#mkdir blib
> > \#cd blib
> > \#unxz perl\-5\.22\.0\.tar\.xz
> > \#tar \-xvf perl\-5\.22\.0\.tar
> > \#cd perl\-5\.22\.0
> > \#\./Configure \-des \-Dprefix=\./
> > \#make
> >
> >
> > 2\. the possible patch is below&#8203;:
> > \+\-\-\-
> > \+ dist/PathTools/Makefile\.PL | 3 \-\-\-
> > \+ 1 file changed\, 3 deletions\(\-\)
> > \+
> > \+diff \-\-git a/dist/PathTools/Makefile\.PL b/dist/PathTools/Makefile\.PL
> > \+index 1b21de4\.\.f562cb2 100644
> > \+\-\-\- a/dist/PathTools/Makefile\.PL
> > \+\+\+\+ b/dist/PathTools/Makefile\.PL
> > \+@&#8203;@&#8203; \-1\,6 \+1\,3 @&#8203;@&#8203;
> > \+\-
> > \+\-BEGIN \{ @&#8203;INC = grep \{\!/blib/\} @&#8203;INC \}
> > \+\-
> > \+ require 5\.005;
> > \+ use ExtUtils&#8203;::MakeMaker;
> > \+ WriteMakefile
> >  \+\-\-
> > \+1\.9\.1
> 
> PathTools is a library which is distributed with the Perl 5 core\, but
> also available on CPAN \(http&#8203;://search\.cpan\.org/dist/PathTools/\)\.  The
> fact that it is in the 'dist/' means that it is primarily maintained
> by the Perl 5 Porters\.
> 
> However\, any deficiency in dist/PathTools/Makefile\.PL would \*at most\*
> affect the compilation of the PathTools library\.  It would \*not\*
> affect the compilation of perl 5 as a whole\.  So I doubt that a change
> to that Makefile\.PL would be the remedy to your problem \-\- if indeed
> your problem is a failure of perl5 as a whole?
> 
> Could you please post\, as an attachment to this bug report \(or email
> attachment\)\, the complete arguments you used to configure Perl and all
> the output of 'sh \./Configure' and 'make' up to the point where you
> experienced a failure?
> 
> Also\, please let us know your OS; how you are attempting to build Perl
> 5; etc\.
> 
> Thank you very much\.

I am using Ubuntu 14\.04 LTS;

the configure log\, make log\, and config\.h are attached

thanks\.

**Comment by p5pRT** [comment]
### From rongqing.li@windriver.com

[config-result](https://rt-archive.perl.org/perl5/Ticket/Attachment/1357124/725931/config-result)

**Comment by p5pRT** [comment]
### From rongqing.li@windriver.com

[config.h](https://rt-archive.perl.org/perl5/Ticket/Attachment/1357124/725932/config.h)

**Comment by p5pRT** [comment]
### From rongqing.li@windriver.com

[make-result](https://rt-archive.perl.org/perl5/Ticket/Attachment/1357124/725933/make-result)

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Jul 12 19&#8203;:08&#8203;:58 2015\, rongqing\.li@&#8203;windriver\.com wrote&#8203;:
> 1\. failed to compile perl5  when path includes the string "blib"
> Reproduce as below&#8203;:
> 
> \#mkdir blib
> \#cd blib
> \#unxz perl\-5\.22\.0\.tar\.xz
> \#tar \-xvf perl\-5\.22\.0\.tar
> \#cd perl\-5\.22\.0
> \#\./Configure \-des \-Dprefix=\./
> \#make
> 

Is there some reason why you are trying to configure and build perl in a directory whose path ends in 'blib'?

For most Perl modules 'blib' is a directory reserved for the build process\.  It is something typically created by 'make'\.  While I suppose there is no law against working in a directory called 'blib'\, it does seem to be asking for trouble\.

Can you explain why you are attempting this?

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From rehsack@gmail.com


> Am 15\.07\.2015 um 04&#8203;:20 schrieb Rongqing Li \<rongqing\.li@&#8203;windriver\.com>&#8203;:
> 
> 
> 
> On 2015年07月15日 09&#8203;:58\, James E Keenan via RT wrote&#8203;:
>> On Sun Jul 12 19&#8203;:08&#8203;:58 2015\, rongqing\.li@&#8203;windriver\.com wrote&#8203;:
>>> 1\. failed to compile perl5  when path includes the string "blib"
>>> Reproduce as below&#8203;:
>>> 
>>> \#mkdir blib
>>> \#cd blib
>>> \#unxz perl\-5\.22\.0\.tar\.xz
>>> \#tar \-xvf perl\-5\.22\.0\.tar
>>> \#cd perl\-5\.22\.0
>>> \#\./Configure \-des \-Dprefix=\./
>>> \#make
>>> 
>> 
>> Is there some reason why you are trying to configure and build perl in a directory whose path ends in 'blib'?
>> 
> 
> no special reason;
> 
> I are building a Yocto project\, which includes perl;
> but I can not building this project under a directory whose
> path has "blib"\, since perl will fail to be compiled
> 
> I think others must spend lots of effect to find why
> compilation failed once he meet this issue\.

As far as I understand\, Roy is complaining about reserved words in path\.

From non\-perl developers point of view\, failing in configure stage with an
error of a reserved word is found in path would make it clearer for anyone\,
but preferred would be a rewrite of the filter only removing the really
issued path components \(I expect the current module build area\)\.

blib already computes them\, maybe sharing them \(our\) and PathTools filter
only the desired part?

I don't know about any other project failing because of reserved word in
build path\.

Cheers
\-\- 
Jens Rehsack \- rehsack@&#8203;gmail\.com

**Comment by p5pRT** [comment]
### From rongqing.li@windriver.com



On 2015年07月15日 09&#8203;:58\, James E Keenan via RT wrote&#8203;:
> On Sun Jul 12 19&#8203;:08&#8203;:58 2015\, rongqing\.li@&#8203;windriver\.com wrote&#8203;:
>> 1\. failed to compile perl5  when path includes the string "blib"
>> Reproduce as below&#8203;:
>>
>> \#mkdir blib
>> \#cd blib
>> \#unxz perl\-5\.22\.0\.tar\.xz
>> \#tar \-xvf perl\-5\.22\.0\.tar
>> \#cd perl\-5\.22\.0
>> \#\./Configure \-des \-Dprefix=\./
>> \#make
>>
>
> Is there some reason why you are trying to configure and build perl in a directory whose path ends in 'blib'?
>

no special reason;

I are building a Yocto project\, which includes perl;
but I can not building this project under a directory whose
path has "blib"\, since perl will fail to be compiled

I think others must spend lots of effect to find why
compilation failed once he meet this issue\.


\-Roy

> For most Perl modules 'blib' is a directory reserved for the build process\.  It is something typically created by 'make'\.  While I suppose there is no law against working in a directory called 'blib'\, it does seem to be asking for trouble\.
>
> Can you explain why you are attempting this?
>
> Thank you very much\.
>

\-\- 
Best Reagrds\,
Roy | RongQing Li

**Comment by p5pRT** [comment]
### From rongqing.li@windriver.com



On 2015年07月15日 13&#8203;:58\, Jens Rehsack via RT wrote&#8203;:
> I don't know about any other project failing because of reserved word in
> build path\.

Yes\, I agree

\-Roy

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Jul 15\, 2015 at 02&#8203;:59&#8203;:36PM \+0800\, Rongqing Li wrote&#8203;:
> 
> 
> On 2015年07月15日 13&#8203;:58\, Jens Rehsack via RT wrote&#8203;:
> >I don't know about any other project failing because of reserved word in
> >build path\.
> 
> Yes\, I agree
> 
> \-Roy

I agree that perl should build regardless of whether the 4 chars "blib"
are in the build directory path or not; I also agree that removing this
line from Makefile\.PL&#8203;:

&nbsp;    BEGIN \{ @&#8203;INC = grep \{\!/blib/\} @&#8203;INC \}

will fix the immediate problem\. However\, since I don't understand what
purpose that line serves / served\, I'd be reluctant to remove it\.

I would speculate that its trying to avoid picking up files from the blib/
subdir in the module's directory\, which presumably may contain rubbish
from a failed earlier build attempt\. In which case maybe the grep line
should be changed to 

&nbsp;    BEGIN \{ @&#8203;INC = grep \{\!/^blib/\} @&#8203;INC \}

? By why that would have been thought necessary\, I don't know\.

The history if this line is that originally PathTools\, as part of Cwd\,
was in perl core \(ext/\) and didn't have this line\. At some point it was
promoted / moved to cpan/ and used an auto\-generated Makefile\.PL\. Then
at some point \(5\.19\.3\)\, the auto\-generated file was replaced with the real
one from the CPAN distribution\, which had\, at some point\, for some reason\,
gained that line\.

\-\- 
In England there is a special word which means the last sunshine
of the summer\. That word is "spring"\.

**Comment by p5pRT** [comment]
### From rehsack@gmail.com



> Am 16\.07\.2015 um 15&#8203;:06 schrieb Dave Mitchell via RT \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> 
> On Wed\, Jul 15\, 2015 at 02&#8203;:59&#8203;:36PM \+0800\, Rongqing Li wrote&#8203;:
>> 
>> 
>> On 2015年07月15日 13&#8203;:58\, Jens Rehsack via RT wrote&#8203;:
>>> I don't know about any other project failing because of reserved word in
>>> build path\.
>> 
>> Yes\, I agree
>> 
>> \-Roy
> 
> I agree that perl should build regardless of whether the 4 chars "blib"
> are in the build directory path or not; I also agree that removing this
> line from Makefile\.PL&#8203;:
> 
>    BEGIN \{ @&#8203;INC = grep \{\!/blib/\} @&#8203;INC \}
> 
> will fix the immediate problem\. However\, since I don't understand what
> purpose that line serves / served\, I'd be reluctant to remove it\.

That's why I voted against that patch in open\-embedded and pressured for
an RT ;\)

It is very reasonable to find the root cause as you started below\.

> I would speculate that its trying to avoid picking up files from the blib/
> subdir in the module's directory\, which presumably may contain rubbish
> from a failed earlier build attempt\. In which case maybe the grep line
> should be changed to 
> 
>    BEGIN \{ @&#8203;INC = grep \{\!/^blib/\} @&#8203;INC \}
> 
> ? By why that would have been thought necessary\, I don't know\.
> 
> The history if this line is that originally PathTools\, as part of Cwd\,
> was in perl core \(ext/\) and didn't have this line\. At some point it was
> promoted / moved to cpan/ and used an auto\-generated Makefile\.PL\. Then
> at some point \(5\.19\.3\)\, the auto\-generated file was replaced with the real
> one from the CPAN distribution\, which had\, at some point\, for some reason\,
> gained that line\.

Maybe Steffen Müller has an idea \(especially at PathTools I'm always
confused whether it's first updated in core and than mirrored to 2nd
life on CPAN or vice versa\)\, or Leon Timmermans or David Golden \(who
was responsible for Module&#8203;::Build&#8203;::Compat version 0\.2808\_01? Oh my\, it
was Ken Williams \- maybe one should ask him  or simply regenerate \.\.\.\)

Cheers
\-\- 
Jens Rehsack \- rehsack@&#8203;gmail\.com

**Comment by p5pRT** [comment]
### From @tux

On Thu\, 16 Jul 2015 14&#8203;:05&#8203;:21 \+0100\, Dave Mitchell \<davem@&#8203;iabyn\.com>
wrote&#8203;:

> On Wed\, Jul 15\, 2015 at 02&#8203;:59&#8203;:36PM \+0800\, Rongqing Li wrote&#8203;:
> > 
> > 
> > On 2015年07月15日 13&#8203;:58\, Jens Rehsack via RT wrote&#8203;:
> > >I don't know about any other project failing because of reserved word in
> > >build path\.
> > 
> > Yes\, I agree
> > 
> > \-Roy
> 
> I agree that perl should build regardless of whether the 4 chars "blib"
> are in the build directory path or not; I also agree that removing this
> line from Makefile\.PL&#8203;:
> 
>     BEGIN \{ @&#8203;INC = grep \{\!/blib/\} @&#8203;INC \}

That is wrong anyway\, it should most likely be /\\bblib\\b/\. I see no
reason whatsoever to skipp "/foo/blibber/duh" or "/foo/foblib/ong"

> will fix the immediate problem\. However\, since I don't understand what
> purpose that line serves / served\, I'd be reluctant to remove it\.
> 
> I would speculate that its trying to avoid picking up files from the blib/
> subdir in the module's directory\, which presumably may contain rubbish
> from a failed earlier build attempt\. In which case maybe the grep line
> should be changed to 
> 
>     BEGIN \{ @&#8203;INC = grep \{\!/^blib/\} @&#8203;INC \}
> 
> ? By why that would have been thought necessary\, I don't know\.
> 
> The history if this line is that originally PathTools\, as part of Cwd\,
> was in perl core \(ext/\) and didn't have this line\. At some point it was
> promoted / moved to cpan/ and used an auto\-generated Makefile\.PL\. Then
> at some point \(5\.19\.3\)\, the auto\-generated file was replaced with the real
> one from the CPAN distribution\, which had\, at some point\, for some reason\,
> gained that line\.
> 


\-\- 
H\.Merijn Brand  http&#8203;://tux\.nl   Perl Monger  http&#8203;://amsterdam\.pm\.org/
using perl5\.00307 \.\. 5\.21   porting perl5 on HP\-UX\, AIX\, and openSUSE
http&#8203;://mirrors\.develooper\.com/hpux/        http&#8203;://www\.test\-smoke\.org/
http&#8203;://qa\.perl\.org   http&#8203;://www\.goldmark\.org/jeff/stupid\-disclaimers/

**Comment by p5pRT** [comment]
### From rongqing.li@windriver.com



On 2015年07月21日 20&#8203;:55\, H\. Merijn Brand via RT wrote&#8203;:
> On Thu\, 16 Jul 2015 14&#8203;:05&#8203;:21 \+0100\, Dave Mitchell \<davem@&#8203;iabyn\.com>
> wrote&#8203;:
>
>> On Wed\, Jul 15\, 2015 at 02&#8203;:59&#8203;:36PM \+0800\, Rongqing Li wrote&#8203;:
>>>
>>>
>>> On 2015年07月15日 13&#8203;:58\, Jens Rehsack via RT wrote&#8203;:
>>>> I don't know about any other project failing because of reserved word in
>>>> build path\.
>>>
>>> Yes\, I agree
>>>
>>> \-Roy
>>
>> I agree that perl should build regardless of whether the 4 chars "blib"
>> are in the build directory path or not; I also agree that removing this
>> line from Makefile\.PL&#8203;:
>>
>>      BEGIN \{ @&#8203;INC = grep \{\!/blib/\} @&#8203;INC \}
>
> That is wrong anyway\, it should most likely be /\\bblib\\b/\. I see no
> reason whatsoever to skipp "/foo/blibber/duh" or "/foo/foblib/ong"

could you submit a patch for this issue

thanks

\-Roy

**Comment by p5pRT** [comment]
### From rongqing.li@windriver.com



On 2015年07月21日 20&#8203;:55\, H\. Merijn Brand via RT wrote&#8203;:
> On Thu\, 16 Jul 2015 14&#8203;:05&#8203;:21 \+0100\, Dave Mitchell \<davem@&#8203;iabyn\.com>
> wrote&#8203;:
>
>> On Wed\, Jul 15\, 2015 at 02&#8203;:59&#8203;:36PM \+0800\, Rongqing Li wrote&#8203;:
>>>
>>>
>>> On 2015年07月15日 13&#8203;:58\, Jens Rehsack via RT wrote&#8203;:
>>>> I don't know about any other project failing because of reserved word in
>>>> build path\.
>>>
>>> Yes\, I agree
>>>
>>> \-Roy
>>
>> I agree that perl should build regardless of whether the 4 chars "blib"
>> are in the build directory path or not; I also agree that removing this
>> line from Makefile\.PL&#8203;:
>>
>>      BEGIN \{ @&#8203;INC = grep \{\!/blib/\} @&#8203;INC \}
>
> That is wrong anyway\, it should most likely be /\\bblib\\b/\. I see no
> reason whatsoever to skipp "/foo/blibber/duh" or "/foo/foblib/ong"
>

it can not fix all issues to replace /blib/ with /\\bblib\\b/\.


/\\bblib\\b/ will make the building path "/tmp/xxxxblibyy/" work\.
but if the building path has sub\-dir blib\, it does not work\, like
/tmp/blib/

since \\b is anchor in perl

\-Roy

**Comment by p5pRT** [comment]
### From rehsack@gmail.com


> Am 29\.07\.2015 um 05&#8203;:05 schrieb rongqing via RT \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> 
> On 2015年07月21日 20&#8203;:55\, H\. Merijn Brand via RT wrote&#8203;:
>> On Thu\, 16 Jul 2015 14&#8203;:05&#8203;:21 \+0100\, Dave Mitchell \<davem@&#8203;iabyn\.com>
>> wrote&#8203;:
>> 
>>> On Wed\, Jul 15\, 2015 at 02&#8203;:59&#8203;:36PM \+0800\, Rongqing Li wrote&#8203;:
>>>> 
>>>> 
>>>> On 2015年07月15日 13&#8203;:58\, Jens Rehsack via RT wrote&#8203;:
>>>>> I don't know about any other project failing because of reserved word in
>>>>> build path\.
>>>> 
>>>> Yes\, I agree
>>>> 
>>>> \-Roy
>>> 
>>> I agree that perl should build regardless of whether the 4 chars "blib"
>>> are in the build directory path or not; I also agree that removing this
>>> line from Makefile\.PL&#8203;:
>>> 
>>>     BEGIN \{ @&#8203;INC = grep \{\!/blib/\} @&#8203;INC \}
>> 
>> That is wrong anyway\, it should most likely be /\\bblib\\b/\. I see no
>> reason whatsoever to skipp "/foo/blibber/duh" or "/foo/foblib/ong"
>> 
> 
> it can not fix all issues to replace /blib/ with /\\bblib\\b/\.
> 
> 
> /\\bblib\\b/ will make the building path "/tmp/xxxxblibyy/" work\.
> but if the building path has sub\-dir blib\, it does not work\, like
> /tmp/blib/
> 
> since \\b is anchor in perl

The more I think about the source of the line \(Module&#8203;::Build&#8203;::Compat\)\,
I vote for completely remove that line and review such imports in future\.

Cheers
\-\- 
Jens Rehsack \- rehsack@&#8203;gmail\.com

**Comment by p5pRT** [comment]
### From @tonycoz

On Thu Jul 16 06&#8203;:18&#8203;:09 2015\, rehsack@&#8203;gmail\.com wrote&#8203;:
> Maybe Steffen Müller has an idea \(especially at PathTools I'm always
> confused whether it's first updated in core and than mirrored to 2nd
> life on CPAN or vice versa\)\, or Leon Timmermans or David Golden \(who
> was responsible for Module&#8203;::Build&#8203;::Compat version 0\.2808\_01? Oh my\, it
> was Ken Williams \- maybe one should ask him  or simply regenerate \.\.\.\)

The offending line was added to fix this bug&#8203;:

https://rt.cpan.org/Ticket/Display.html?id=4681

The line wasn't generated by Module&#8203;::Build&#8203;::Compat

Tony

**Comment by p5pRT** [comment]
### From rehsack@gmail.com


> Am 12\.08\.2015 um 08&#8203;:41 schrieb Tony Cook via RT \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> 
> On Thu Jul 16 06&#8203;:18&#8203;:09 2015\, rehsack@&#8203;gmail\.com wrote&#8203;:
>> Maybe Steffen Müller has an idea \(especially at PathTools I'm always
>> confused whether it's first updated in core and than mirrored to 2nd
>> life on CPAN or vice versa\)\, or Leon Timmermans or David Golden \(who
>> was responsible for Module&#8203;::Build&#8203;::Compat version 0\.2808\_01? Oh my\, it
>> was Ken Williams \- maybe one should ask him  or simply regenerate \.\.\.\)
> 
> The offending line was added to fix this bug&#8203;:
> 
> https://rt.cpan.org/Ticket/Display.html?id=4681

The bug looks very weird \- where comes the dirty Cwd 2\.08 stuff from?
I'm not sure the line fixes the root cause\, it seems to work on symptom\.\.\.

However \- 10 years later searching and fixing the root\-cause might not
be reasonable at all\.

Let's try to figure out how to fix the symptom fix to avoid breakage
of third party environments&#8203;:

Merijn proposed using /\\bblib\\b/ instead of /blib/ \- looking at rt\#4681
I would suggest using /^\\bblib\\b/

That should be restrictive enough to avoid disturbing anything else than
dirty blib/inc \.\.\.

> The line wasn't generated by Module&#8203;::Build&#8203;::Compat

Thanks for investigating\, Tony\.

Best regards
\-\- 
Jens Rehsack \- rehsack@&#8203;gmail\.com

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Aug 12\, 2015 at 03&#8203;:13&#8203;:38PM \+0200\, Jens Rehsack wrote&#8203;:
> 
> > Am 12\.08\.2015 um 08&#8203;:41 schrieb Tony Cook via RT \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> > 
> > On Thu Jul 16 06&#8203;:18&#8203;:09 2015\, rehsack@&#8203;gmail\.com wrote&#8203;:
> >> Maybe Steffen Müller has an idea \(especially at PathTools I'm always
> >> confused whether it's first updated in core and than mirrored to 2nd
> >> life on CPAN or vice versa\)\, or Leon Timmermans or David Golden \(who
> >> was responsible for Module&#8203;::Build&#8203;::Compat version 0\.2808\_01? Oh my\, it
> >> was Ken Williams \- maybe one should ask him  or simply regenerate \.\.\.\)
> > 
> > The offending line was added to fix this bug&#8203;:
> > 
> > https://rt.cpan.org/Ticket/Display.html?id=4681
> 
> The bug looks very weird \- where comes the dirty Cwd 2\.08 stuff from?
> I'm not sure the line fixes the root cause\, it seems to work on symptom\.\.\.
> 
> However \- 10 years later searching and fixing the root\-cause might not
> be reasonable at all\.
> 
> Let's try to figure out how to fix the symptom fix to avoid breakage
> of third party environments&#8203;:
> 
> Merijn proposed using /\\bblib\\b/ instead of /blib/ \- looking at rt\#4681
> I would suggest using /^\\bblib\\b/
> 
> That should be restrictive enough to avoid disturbing anything else than
> dirty blib/inc \.\.\.
> 
> > The line wasn't generated by Module&#8203;::Build&#8203;::Compat
> 
> Thanks for investigating\, Tony\.

I've just applied the following to blead\. Hopefully its a suitable
compromise\.

commit 850e14d313dc5493e5162c407f21ae0a9fec2805
Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
AuthorDate&#8203;: Tue Sep 27 16&#8203;:43&#8203;:30 2016 \+0100
Commit&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
CommitDate&#8203;: Tue Sep 27 16&#8203;:48&#8203;:53 2016 \+0100

&nbsp;    Cwd\.xs&#8203;: avoid blib better while building
&nbsp;    
&nbsp;    RT \# 125603
&nbsp;    
&nbsp;    There's an old line in Cwd's Makefile\.PL&#8203;:
&nbsp;    
&nbsp;        BEGIN \{ @&#8203;INC = grep \{\!/blib/\} @&#8203;INC \}
&nbsp;    
&nbsp;    This was added 12 years ago to solve a problem with a static perl and
&nbsp;    building a newer Cwd \(but no\-one's quite sure what \*exactly\* the issue was
&nbsp;    any more\)\.
&nbsp;    
&nbsp;    However\, this breaks building perl under a directory that has 'blib'
&nbsp;    in the pathname\. This commit estricts the grep  to just exclusing exactly
&nbsp;    blib/lib and blib/arch\.
&nbsp;    
&nbsp;    This should hopefully still workaround the static build issue\, while no
&nbsp;    longer breaking perl builds\.


Affected files \.\.\.
&nbsp;    M    dist/PathTools/Makefile\.PL

Differences \.\.\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/dist/PathTools/Makefile.PL b/dist/PathTools/Makefile.PL
index 6cd70fa..71b9a60 100644
--- a/dist/PathTools/Makefile.PL
+++ b/dist/PathTools/Makefile.PL
@@ -1,5 +1,11 @@
 
-BEGIN { @INC = grep {!/blib/} @INC }
+# See https://rt.cpan.org/Public/Bug/Display.html?id=4681
+# and https://rt.perl.org/Ticket/Display.html?id=125603
+# When installing a newer Cwd on a system with an existing Cwd,
+# under some circumstances the old Cwd.pm and the new Cwd.xs could
+# get mixed up and SEGVs ensue.
+
+BEGIN { @INC = grep { $_ ne "blib/arch" and $_ ne "blib/lib" } @INC }


-- 
```
</details>

"Foul and greedy Dwarf \- you have eaten the last candle\."
&nbsp;    \-\- "Hordes of the Things"\, BBC Radio\.

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
